### PR TITLE
Remove dynamic invocation

### PR DESCRIPTION
**Pull request by rrousselGit**
Since the matchers are using FeatureMatcher, the typedMatches method receives a typed object instead of dynamic

### CODE REVIEW

**Comment by jakemac53** [comment]
LGTM, can  you bump this to 0.12.15-dev and add a changelog entry? You can leave the message blank, since its not a user facing change.

**Review by jakemac53 — APPROVED** [review]

**Comment by rrousselGit** [comment]
Ah sorry, I missed your previous message. Looks like the changelog entry + bump has been done already

File: lib/src/core_matchers.dart [hunk 1]
```diff
@@ -14,7 +14,9 @@ class _Empty extends Matcher {
   const _Empty();
 
   @override
-  bool matches(Object? item, Map matchState) => (item as dynamic).isEmpty;
+  bool matches(Object? item, Map matchState) =>
+      // ignore: avoid_dynamic_calls
```
Comment (natebosch) [inline lib/src/core_matchers.dart:18]: It's too bad we don't have a better way to suppress these than the ignore. I filed https://github.com/dart-lang/linter/issues/4806

File: lib/src/expect/throws_matcher.dart [hunk 2]
```diff
@@ -73,7 +73,7 @@ class Throws extends AsyncMatcher {
   // function.
   @override
   dynamic /*FutureOr<String>*/ matchAsync(Object? item) {
-    if (item is! Function && item is! Future) {
+    if (item is! Function() && item is! Future) {
```
Comment (natebosch) [inline lib/src/expect/throws_matcher.dart:76]: This stopped accepting values which are a `Function<T>()`. I think we may need to keep this as a dynamic call to handle generic methods. This is why the `Future.value` tear-off case is failing.

cc @jakemac53 - I have completely overlooked the fact that functions are not subtypes if they have different numbers of generic type arguments... I wonder if there are any other places we have been more strict than we should and are unintentionally blocking generic functions used as callbacks.

File: lib/src/core_matchers.dart [hunk 3]
```diff
@@ -144,11 +148,11 @@ class isInstanceOf<T> extends TypeMatcher<T> {
 /// a wrapper will have to be created.
 const Matcher returnsNormally = _ReturnsNormally();
 
-class _ReturnsNormally extends FeatureMatcher<Function> {
+class _ReturnsNormally extends FeatureMatcher<Function()> {
   const _ReturnsNormally();
 
   @override
-  bool typedMatches(Function f, Map matchState) {
+  bool typedMatches(Function() f, Map matchState) {
```
Comment (natebosch) [inline lib/src/core_matchers.dart:155]: This is likely a behavior change that we don't have tests for...

File: lib/src/core_matchers.dart (hunk 1 above)
Comment (rrousselGit) [reply lib/src/core_matchers.dart:18]: Could this be replaced with `if (item is Iterable) return item.isEmpty; else if (item is String) return item.isEmpty;`? 

That would technically be breaking if someone defined an unknown object with an `isEmpty` getter. But the fact that it currently works is surprising to me. I thought the implementation relied on pattern match, not dynamic invocation

Comment (natebosch) [reply lib/src/core_matchers.dart:18]: I think supporting an `isEmpty` property on an arbitrary user defined class was an intentional feature and I'm not eager to try to remove it. It's not a design we'd carry forward to `checks`.

**Pull request closed without merging**